### PR TITLE
ASR: update from LFortran

### PR DESCRIPTION
**Pull request by certik**
Corresponding LFortran PR: https://github.com/lfortran/lfortran/pull/73

### CODE REVIEW

File: grammar/asdl_cpp.py [hunk 1]
```diff
@@ -915,94 +1038,94 @@ def visitField(self, field):
...
-                self.emit("    current_stmt_copy = current_stmt;", level)
-                self.emit("    current_stmt = &(x->m_%s[i]);" % (field.name), level)
-                self.emit("    self().replace_stmt(x->m_%s[i]);"%(field.name), level)
-                self.emit("    current_stmt = current_stmt_copy;", level)
-                self.emit("}", level)
+     def __init__(self, stream, data):
+         self.replace_stmt = []
+         self.is_stmt = False
+         self.is_product = False
+         super(StmtBaseReplacerVisitor, self).__init__(stream, data)
```
Comment (certik) [inline grammar/asdl_cpp.py:1045]: All these changes use 5 spaces instead of 4. :(

**Comment by certik** [comment]
@czgdp1807 let me know if you see any issues with any of these changes. This might be easier to review & merge than the corresponding LFortran PR. This brings it a lot closer to LFortran. If it looks good, I'll probably just squash the changes into one commit.

**Comment by certik** [comment]
@namannimmo10, @Smit-create do you see any issues with these changes? These bring LPython much closer to LFortran, but I want to make sure I didn't inadvertently revert some change in LPython that we want.

**Comment by czgdp1807** [comment]
Well, I am working on LFortran's fixes in https://github.com/lfortran/lfortran/pull/73/files. So, will push some changes here as well to keep everything in sync before we merge this.

**Comment by certik** [comment]
That works. I think these two PRs are *almost* in sync. There are a few more changes from LPython that didn't make it to https://github.com/lfortran/lfortran/pull/73 because they completely broke the tests, but I think I just ported them incorrectly.

**Comment by czgdp1807** [comment]
@certik The tests pass here.

**Comment by certik** [comment]
Awesome, thank you @czgdp1807 ! I am going to squash these into just one commit and merge.

**Comment by Smit-create** [comment]
Thank you!!

**Pull request merged**